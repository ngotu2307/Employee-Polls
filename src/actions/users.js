export const RECEIVE_USERS = "RECEIVE_USERS";
export const SAVE_USER_ANSWER = "SAVE_USER_ANSWER";
export const SAVE_NEW_QUESTION_USER = "SAVE_NEW_QUESTION_USER";

export function receiveUsers(users) {
    return {
        type: RECEIVE_USERS,
        users,
    }
}

export function saveUserAnswer({ authedUser, qid, answer }) {
    return {
        type: SAVE_USER_ANSWER,
        authedUser,
        qid,
        answer,
    };
}

export function saveNewQuestionUser({author, id}) {
    return {
        type: SAVE_NEW_QUESTION_USER,
        author,
        qid: id,
    };
}