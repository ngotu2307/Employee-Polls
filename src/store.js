import { createStore } from "redux";
import persistReducer from "./reducers";
import middleware from "./middleware";

export const store = createStore(persistReducer, middleware);