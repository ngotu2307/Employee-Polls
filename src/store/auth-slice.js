import { createSlice, configureStore } from "@reduxjs/toolkit";
const initialAuthState = { isAuthen: false }

const authSlice = createSlice({
    name: 'auth',
    initialState: initialAuthState,
    reducers: {
        login(state) {
            state.isAuthen = true
        },
        logout(state) {
            state.isAuthen = false
        }
    }
})

export const authActions = authSlice.actions

export default authSlice.reducer;