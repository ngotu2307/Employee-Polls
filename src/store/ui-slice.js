import { createSlice, configureStore } from "@reduxjs/toolkit";
const initialUiState = { isAuthen: true }

const uiSlice = createSlice({
    name: 'auth',
    initialState: initialUiState,
    reducers: {
        login(state) {
            state.isAuthen = true
        },
        logout(state) {
            state.isAuthen = false
        }
    }
})

export const uiActions = uiSlice.actions

export default uiSlice.reducer;