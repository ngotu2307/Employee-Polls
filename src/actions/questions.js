import { saveQuestionAnswerAPI, saveNewQuestionAPI } from "../utils/api";
import { saveUserAnswer, saveNewQuestionUser } from "./users";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const SAVE_QUESTION_ANSWER = "SAVE_QUESTION_ANSWER";
export const SAVE_NEW_QUESTION = "SAVE_NEW_QUESTION";

export function receiveQuestions(questions) {
    return {
        type: RECEIVE_QUESTIONS,
        questions,
    }
}

export function saveQuestionAnswer({ authedUser, qid, answer }) {
    return {
        type: SAVE_QUESTION_ANSWER,
        authedUser,
        qid,
        answer
    }
}

export function saveNewQuestion(question) {
    return {
        type: SAVE_NEW_QUESTION,
        question
    }
}

export function handerSaveQuestionAnswer(info) {
    return (dispatch) => {
        return saveQuestionAnswerAPI(info)
            .then(() => {
                dispatch(saveQuestionAnswer(info));
                dispatch(saveUserAnswer(info));
            })
            .catch((e) => {
                console.warn("Error in saveQuestionAnswerAPI: ", e);
                alert("There was an error when saving question answer!");
            })
    }
}

export function handerSaveNewQuestion(info) {
    return (dispatch) => {
        return saveNewQuestionAPI(info)
            .then((question) => {
                dispatch(saveNewQuestion(question));
                dispatch(saveNewQuestionUser(question));
            })
            .catch((e) => {
                console.warn("Error in saveNewQuestionAPI: ", e);
                alert("There was an error when saving new quesion!");
            })
    }
}