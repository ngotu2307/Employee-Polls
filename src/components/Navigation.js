import { Link } from 'react-router-dom';
import { connect } from "react-redux";
import { useNavigate, useLocation } from 'react-router-dom';
import { setAuthedUser } from '../actions/authedUser';

const Navigation = ({ dispatch, user, authedUser }) => {
  const navigate = useNavigate();
  const location = useLocation();
  console.log("Location: ", location.pathname);

  const handleClickLogout = (e) => {
    console.log('Logout!');
    e.preventDefault();
    dispatch(setAuthedUser(null));
    navigate("/");
  }

  console.log("user: ", user);

  return (
    <div>
      <div className='d-flex justify-content-between align-items-center mt-1'>
        <ul className='list-group list-group-horizontal'>
          <li className="list-group-item border-0">
            <Link style={{textDecoration: location.pathname==="/dashboard" ? "underline" : "none"}} to="/">Home</Link>
          </li>
          <li className="list-group-item border-0">
            <Link style={{textDecoration: location.pathname==="/leaderboard" ? "underline" : 'none'}} to="/leaderboard">Leaderboard</Link>
          </li>
          <li className="list-group-item border-0">
            <Link style={{textDecoration: location.pathname==="/newpoll" ? "underline" : 'none'}} to="/newpoll">New</Link>
          </li>
        </ul>
        <ul className='list-group list-group-horizontal'>
          <li className="list-group-item border-0">
            <div className='d-flex justify-content-end'>
              {/*need to check user !== undefined in case of refresh browser*/}
              <img width="10%" src={user !== undefined ? user.avatarURL : undefined} className='rounded-circle' alt="Profile" />
              <span data-testid="authedUser-text" className='ms-2 fw-bold'>{authedUser}</span>
            </div>
          </li>
          <li className="list-group-item border-0">
            <Link style={{textDecoration: 'none'}} onClick={handleClickLogout}>Logout</Link>
          </li>
        </ul>
      </div>

      <div className="border-top mb-3 mt-1"></div>
    </div>
  );
}

const mapStateToProps = ({ authedUser, users }) => {
  const userIds = Object.keys(users);
  const authenUserId = userIds.find(userId => userId === authedUser);
  return ({
    user: users[authenUserId],
    authedUser
  })
}

export default connect(mapStateToProps)(Navigation);