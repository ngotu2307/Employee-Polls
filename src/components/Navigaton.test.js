import { render } from "@testing-library/react";
import React from 'react';
import Navigation from "./Navigation";
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from "react-router-dom";
import { store, persistor } from "../store";
import { setAuthedUser } from "../actions/authedUser";
import '@testing-library/jest-dom'
import { PersistGate } from 'redux-persist/integration/react'

describe('Navigation', () => {
    it("should render the component", () => {
        var component = render(
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistor}>
                    <Router>
                        <Navigation />
                    </Router>
                </PersistGate>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it("will show authed username", () => {
        store.dispatch(setAuthedUser("sarahedo"));

        const component = render(
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistor}>
                    <Router>
                        <Navigation />
                    </Router>
                </PersistGate>
            </Provider>
        );
        expect(component.getByTestId('authedUser-text')).toHaveTextContent("sarahedo");
    });
});