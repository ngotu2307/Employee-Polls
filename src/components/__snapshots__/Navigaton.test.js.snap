// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="d-flex justify-content-between align-items-center mt-1"
        >
          <ul
            class="list-group list-group-horizontal"
          >
            <li
              class="list-group-item border-0"
            >
              <a
                href="/"
                style="text-decoration: none;"
              >
                Home
              </a>
            </li>
            <li
              class="list-group-item border-0"
            >
              <a
                href="/leaderboard"
                style="text-decoration: none;"
              >
                Leaderboard
              </a>
            </li>
            <li
              class="list-group-item border-0"
            >
              <a
                href="/add"
                style="text-decoration: none;"
              >
                New
              </a>
            </li>
          </ul>
          <ul
            class="list-group list-group-horizontal"
          >
            <li
              class="list-group-item border-0"
            >
              <div
                class="d-flex justify-content-end"
              >
                <img
                  alt="Profile"
                  class="rounded-circle"
                  width="10%"
                />
                <span
                  class="ms-2 fw-bold"
                  data-testid="authedUser-text"
                />
              </div>
            </li>
            <li
              class="list-group-item border-0"
            >
              <a
                href="/"
                style="text-decoration: none;"
              >
                Logout
              </a>
            </li>
          </ul>
        </div>
        <div
          class="border-top mb-3 mt-1"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="d-flex justify-content-between align-items-center mt-1"
      >
        <ul
          class="list-group list-group-horizontal"
        >
          <li
            class="list-group-item border-0"
          >
            <a
              href="/"
              style="text-decoration: none;"
            >
              Home
            </a>
          </li>
          <li
            class="list-group-item border-0"
          >
            <a
              href="/leaderboard"
              style="text-decoration: none;"
            >
              Leaderboard
            </a>
          </li>
          <li
            class="list-group-item border-0"
          >
            <a
              href="/add"
              style="text-decoration: none;"
            >
              New
            </a>
          </li>
        </ul>
        <ul
          class="list-group list-group-horizontal"
        >
          <li
            class="list-group-item border-0"
          >
            <div
              class="d-flex justify-content-end"
            >
              <img
                alt="Profile"
                class="rounded-circle"
                width="10%"
              />
              <span
                class="ms-2 fw-bold"
                data-testid="authedUser-text"
              />
            </div>
          </li>
          <li
            class="list-group-item border-0"
          >
            <a
              href="/"
              style="text-decoration: none;"
            >
              Logout
            </a>
          </li>
        </ul>
      </div>
      <div
        class="border-top mb-3 mt-1"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
